server/src/main/java/org/elasticsearch/index/IndexService.java

```diff
@@ -661,6 +668,46 @@ public SearchExecutionContext newSearchExecutionContext(
         );
     }
 
+    /**
+     * Creates a new {@link QueryRewriteContext}.
+     * This class is used to rewrite queries before we are able to get a valid {@link SearchExecutionContext} and
+     * allows us to anticipate rewriting queries before the query is executed on the data node. Not using a full
+     * SearhcExecutionContext allows us to save on query latency and IO operations.
```

Comment:
SearchExecutionContext is mistyped. Shall we also clarify that this is for cases where access to the index is not required as we can make decisions based on mappings alone? Also, expand on what saves query latency and IO operations? e.g. not pulling a searcher removes the cost of pulling a searcher as well as the associated cost of refreshing idle shards.